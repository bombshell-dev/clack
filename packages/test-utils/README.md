# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
