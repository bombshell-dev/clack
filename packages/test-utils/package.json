{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
