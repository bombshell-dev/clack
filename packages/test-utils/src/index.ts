export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
