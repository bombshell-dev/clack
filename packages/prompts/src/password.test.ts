import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
import * as prompts from './index.js';
import { MockReadable, MockWritable } from './test-utils.js';

describe.each(['true', 'false'])('password (isCI = %s)', (isCI) => {
	let originalCI: string | undefined;
	let output: MockWritable;
	let input: MockReadable;

	beforeAll(() => {
		originalCI = process.env.CI;
		process.env.CI = isCI;
	});

	afterAll(() => {
		process.env.CI = originalCI;
	});

	beforeEach(() => {
		output = new MockWritable();
		input = new MockReadable();
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('renders message', async () => {
		const result = prompts.password({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });

		await result;

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders masked value', async () => {
		const result = prompts.password({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', 'x', { name: 'x' });
		input.emit('keypress', 'y', { name: 'y' });
		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe('xy');
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders custom mask', async () => {
		const result = prompts.password({
			message: 'foo',
			mask: '*',
			input,
			output,
		});

		input.emit('keypress', 'x', { name: 'x' });
		input.emit('keypress', 'y', { name: 'y' });
		input.emit('keypress', '', { name: 'return' });

		await result;

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders and clears validation errors', async () => {
		const result = prompts.password({
			message: 'foo',
			validate: (value) => {
				if (value.length < 2) {
					return 'Password must be at least 2 characters';
				}

				return undefined;
			},
			input,
			output,
		});

		input.emit('keypress', 'x', { name: 'x' });
		input.emit('keypress', '', { name: 'return' });
		input.emit('keypress', 'y', { name: 'y' });
		input.emit('keypress', '', { name: 'return' });

		await result;

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders cancelled value', async () => {
		const result = prompts.password({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', 'x', { name: 'x' });
		input.emit('keypress', '', { name: 'escape' });

		const value = await result;

		expect(prompts.isCancel(value)).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});
});
