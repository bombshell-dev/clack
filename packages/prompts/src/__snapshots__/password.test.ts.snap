// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`password (isCI = false) > renders and clears validation errors 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[33mâ–²[39m  foo
[33mâ”‚[39m  â–ª
[33mâ””[39m  [33mPassword must be at least 2 characters[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[36mâ—†[39m  foo
[36mâ”‚[39m  â–ªâ–ª[7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2mâ–ªâ–ª[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = false) > renders cancelled value 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[31mâ– [39m  foo
[90mâ”‚[39m  [9m[2mâ–ª[22m[29m
[90mâ”‚[39m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = false) > renders custom mask 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  *[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  **[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2m**[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = false) > renders masked value 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ªâ–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2mâ–ªâ–ª[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = false) > renders message 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m[2m[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = true) > renders and clears validation errors 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[33mâ–²[39m  foo
[33mâ”‚[39m  â–ª
[33mâ””[39m  [33mPassword must be at least 2 characters[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[36mâ—†[39m  foo
[36mâ”‚[39m  â–ªâ–ª[7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2mâ–ªâ–ª[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = true) > renders cancelled value 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[31mâ– [39m  foo
[90mâ”‚[39m  [9m[2mâ–ª[22m[29m
[90mâ”‚[39m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = true) > renders custom mask 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  *[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  **[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2m**[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = true) > renders masked value 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[2B",
  "[2K[G",
  "[36mâ”‚[39m  â–ªâ–ª[7m[8m_[28m[27m",
  "[2B",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2mâ–ªâ–ª[22m",
  "
",
  "[?25h",
]
`;

exports[`password (isCI = true) > renders message 1`] = `
[
  "[?25l",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "[999D[4A",
  "[1B",
  "[J",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m[2m[22m",
  "
",
  "[?25h",
]
`;
